--- a/net/minecraft/world/level/material/EmptyFluid.java
+++ b/net/minecraft/world/level/material/EmptyFluid.java
@@ -12,7 +_,7 @@
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public class EmptyFluid extends Fluid {
+public class EmptyFluid extends Fluid implements net.neoforged.neoforge.common.extensions.IEmptyFluidExtension {
     @Override
     public Item getBucket() {
         return Items.AIR;
